modules/product/src/Form/AddToCartForm.php

```diff
@@ -166,4 +182,253 @@ function submitForm(array &$form, FormStateInterface $form_state) {
+      $this->attributeElement($form, $data, $data['used_options']);
+    }
+
+    if (!empty($form['attributes'])) {
+      $form['attributes'] += [
+        '#tree' => TRUE,
+        '#prefix' => '<div id="attribute-widgets" class="attribute-widgets">',
+        '#suffix' => '</div>',
+        '#weight' => 0,
+      ];
+      $form['unchanged_attributes'] += [
+        '#tree' => 'TRUE',
+      ];
+    }
+  }
+
+  /**
+   * Adds an attribute form element to add to cart form.
+   *
+   * @param array $form
+   *   An associative array containing the structure of the form.
+   * @param array $attribute
+   *   The attribute information array.
+   * @param $used_options
+   *   An array of options in use.
+   */
+  protected function attributeElement(array &$form, array $attribute, array $used_options) {
+    $field_name = $attribute['name'];
+    $options = array_intersect_key($attribute['options'], array_combine($used_options[$field_name], $used_options[$field_name]));
+    $default_value = $this->defaultVariation->{$field_name}->first()->entity->id();
+
+    $form['attributes'][$field_name] = [
+      '#type' => $attribute['type'],
+      '#title' => $attribute['title'],
+      '#options' => $options,
+      '#default_value' => $default_value,
+      '#ajax' => [
+        'callback' => '::attributesAjax',
+        'wrapper' => 'attribute-widgets',
+      ],
+      '#value' => NULL,
+    ];
+
+    // Add the empty value if the field is not required and products on
+    // the form include the empty value.
+    if (!$attribute['required'] && in_array('', $used_options[$field_name])) {
+      $form['attributes'][$field_name]['#empty_value'] = '';
+    }
+
+    $form['unchanged_attributes'][$field_name] = [
+      '#type' => 'hidden',
+      '#value' => $default_value,
+    ];
+  }
+
+  /**
+   * Gets an array of attributes and their settings.
+   *
+   * @param \Drupal\commerce_product\Entity\ProductVariationInterface[] $variations
+   *   The product variations.
+   *
+   * @return array
+   *   Information about the attributes.
+   */
+  protected function getAttributesInfo(array $variations) {
+    $attributes = [];
+    /** @var \Drupal\Core\Field\FieldConfigInterface $field */
+    foreach ($this->defaultVariation->getAttributeFieldDefinitions() as $field) {
+      $third_party_settings = $field->getThirdPartySettings('commerce_product');
+
+      $attribute_label = !empty($third_party_settings['attribute_widget_title']) ?
+        $third_party_settings['attribute_widget_title'] : $field->label();
+
+      $attributes[$field->getName()] = [
+        'name' => $field->getName(),
+        'type' => $third_party_settings['attribute_widget'],
+        'title' => $attribute_label,
+        'required' => $field->isRequired(),
+        'options' => [],
+      ];
+
+      // Get possible field options based on current variations.
+      foreach ($variations as $variation) {
+        $attribute_value = $variation->{$field->getName()}->first()->entity;
```

Comment:
Just like in the target_id example, the first() here isn't needed, TypedData does that for you.